Add fix storing email links and open tracked links [MAILPOET-6433]

lysyjan opened the pull request:
## Description

When we released version 5.5.2, it contained a fix in rendering links. Nevertheless, we didn't know about another bug in TinyMCE, which can cause doubled encoding & when the link is pasted via `smart_paste`.

- This PR adds a fix for TinyMCE to avoid broken links in stored emails.
- It also adds a change for correct extraction of the broken mails because the broken links can be stored in post notifications, etc.
- A new migration fixes stored incorrect links.

## Code review notes

_N/A_

## QA notes

_N/A_

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-6433]

## After-merge notes

_N/A_

## Tasks

- [ ] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [ ] I added sufficient test coverage
- [ ] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-6433]: https://mailpoet.atlassian.net/browse/MAILPOET-6433?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Preview

[Preview in WordPress Playground](https://account.mailpoet.com/playground/new/branch:email-links-fix)

_The latest successful build from `email-links-fix` will be used. If none is available, the link won't work._

lysyjan committed:
Fix storing link in emails from TinyMCE

[MAILPOET-6433]

lysyjan committed:
Add a fix for rendering incorrect links

[MAILPOET-6433]

lysyjan committed:
Add migration to fix broken links

[MAILPOET-6433]

costasovo commented on mailpoet/lib/Migrations/App/Migration_20250120_094614_App.php:
```diff
@@ -0,0 +1,39 @@
+  public function run(): void {
+    $sendingQueues = $this->getSendingQueues();
+    $linksRepository = $this->container->get(NewsletterLinkRepository::class);
+    foreach ($sendingQueues as $sendingQueue) {
+      $links = $linksRepository->findBy(['queue' => $sendingQueue]);
+      foreach ($links as $link) {
+        $link->setUrl(str_replace('&amp;amp;', '&amp;', $link->getUrl()));
+      }
```
I would consider using just a single raw update query to avoid fetching all the links. I have a concern that we may need to fetch many sending queues, and the sending queue contains the fully rendered email, so there might be memory issues, especially for users who send a lot of automatic emails and/or update later.  This query worked for me:
```sql
UPDATE wp_mailpoet_newsletter_links
SET url = REPLACE(url, '&amp;amp;', '&amp;')
WHERE queue_id > {ID_OF_SENDING_QUEUE_FROM_2024-12-24};
```

lysyjan replied:
Good point. I will update the migration.

alex-mpoet commented on mailpoet/lib/Newsletter/Renderer/PostProcess/OpenTracking.php:
```diff
@@ -36,7 +36,8 @@ private static function appendToDomNodes($template, $openTrackingImage): void {
       array_map(
-        fn($item) => $item->html($item->toString(true, true, 1) . $openTrackingImage),
+        // Because TinyMCE sometimes converts `&amp;` to `&amp;amp;` we need to replace it back because the previous versions contained a bug which rendered the links correctly.
+        fn($item) => $item->html(str_replace('&amp;amp;', '&amp;', $item->toString(true, true, 1)) . $openTrackingImage),
```
This one-liner starts getting quite big, could it be split into a couple of lines?

alex-mpoet commented on mailpoet/lib/Migrations/App/Migration_20250120_094614_App.php:
```diff
@@ -0,0 +1,43 @@
+use MailPoet\Migrator\AppMigration;
+
+/**
+ * The plugin from the version 5.5.2 and higher contained a bug when the `&amp;` in links was stored
```
It would be nice to document the highest affected version too, if possible.

lysyjan committed:
Fix code readability and improve migration comment

[MAILPOET-6433]

lysyjan commented:
@alex-mpoet I tried to address your comments in the last commit.

alex-mpoet reviewed (dismissed):
Thank you for the changes! Looks good!

costasovo committed:
Improve fix for amp in links and move it to renderer

- The fix addresses the issue that we can't have &amp in a link because it breaks it.
- By moving the fix to renderer post-process the fix is applied also for sites
that have tracking disabled
- The fix is now targeted only to href attributes in anchor tags
[MAILPOET-6433]

costasovo committed:
Update migration to decode all encoded ampersands in links

[MAILPOET-6433]

Aschepikov reviewed (approved):
GJ!
